From f31dea8d75f23f5b37ad3c9fa05a44078edceef0 Mon Sep 17 00:00:00 2001
From: mcarlin <mcarlin@datumsystems.com>
Date: Fri, 9 Oct 2020 14:47:22 -0700
Subject: [PATCH 5/6] Fix rtl8211f LED function.

---
 drivers/net/phy/realtek.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/net/phy/realtek.c b/drivers/net/phy/realtek.c
index c76df51dd3c5..81a86ab556fa 100644
--- a/drivers/net/phy/realtek.c
+++ b/drivers/net/phy/realtek.c
@@ -175,6 +175,12 @@ static int rtl8211f_config_init(struct phy_device *phydev)
 	u16 val;
 	int ret;
 
+	/* LED0=10/1000+ACT, LED1=100/1000+ACT, LED2=Disabled */
+	phy_write(phydev, RTL821x_PAGE_SELECT, 0xd04);
+	phy_write(phydev, 0x10, 0x0359);
+	phy_write(phydev, 0x11, 0x0000);
+	phy_write(phydev, RTL821x_PAGE_SELECT, 0x0);
+
 	/* enable TX-delay for rgmii-{id,txid}, and disable it for rgmii and
 	 * rgmii-rxid. The RX-delay can be enabled by the external RXDLY pin.
 	 */
-- 
2.28.0

